### ISSUE TITLE
FireMonkey: Userscript editing and linting

### ISSUE THREAD

**Issue by wltb** [erosman/support#632]
I have a problem that is maybe related to #623: Newer JS syntax is not linted right, e.g. in classes, the private property marker **#** throws the linter off. A small (non-sensical) example:

```javascript
class className {
  #prop;
  static #prop2;
  static {
    this.#prop2 = "txt";
  }
  constructor(arg) {
    this.#prop = arg;
    this.constructor.#func(arg);
  }

  static #func(arg) {
    return arg;
  }

  func2() {
    for(const [foo, bar] of Object.entries(this.#prop)) {
      if(this.constructor.#prop2.has(bar.toLowerCase())) this.constructor.#func(foo);
    }
  }
}
```

Maybe updating the linter solves this? BTW, CodeMirror fixed some small annoyances w.r.t. to drag+move/copy last year, an update that updates CodeMirror would be appreciated in any case.

Finally, CodeMirror apparently can be tweaked to allow autocomplete without pressing _ctrl+space_, see [here](https://discuss.codemirror.net/t/autocomplete-always-show-suggestions/1822) for a small working configuration. Since this involves tweaking events, it seems that it can't work in FireMonkey as of now. Could something like this be added as an option in the future?

**Comment by erosman** [comment]
> I have a problem that is maybe related to #623: Newer JS syntax is not linted right ...

It is similar.
FireMonkey uses [CodeMirror 5](https://github.com/codemirror/codemirror5) for the code editor and CodeMirror 5 uses [JSHint](https://github.com/jshint/jshint) to lint the JavaScript.

Unfortunately, JSHint has been abandoned (last release Nov 11, 2022), therefore it doesn't support newer JS implementations.
See also: 
- [esversion beyond 11](https://github.com/jshint/jshint/issues/3645)

## CodeMirror 6
[CodeMirror 5](https://github.com/codemirror/codemirror5) development has slowed down in favour of [CodeMirror 6](https://github.com/codemirror/dev) which uses [ESLint](https://github.com/eslint/eslint).

In order to use it in a browser extension, CodeMirror has to be compiled and I haven't found a guide for it yet.
I plan to switch to CodeMirror 6 as soon as I find a guide.

See also:
- [Is there a step-by-step guide to pack CodeMirror to use in a browser extension?](https://discuss.codemirror.net/t/is-there-a-step-by-step-guide-to-pack-codemirror-to-use-in-a-browser-extension/8139)
- [Is a browser version of CodeMirror 6 available?](https://discuss.codemirror.net/t/is-a-browser-version-of-codemirror-6-available/2729)

**Comment by wltb** [comment]
Howdy @erosman,

I wish I could help you, but I've never used javascript outside of a browser. The only helpful thing I found was this [small tutorial](https://codemirror.net/examples/bundle/) about bundling _codemirror6_ using _rollup_, but one first has to write a JS file and include the needed CM modules there oneself.

Maybe that helps a little?

**Comment by erosman** [comment]
Sadly, the node.js bundling software create a massive code. It is not only CodeMirror but Eslint & StyleLint that have to be packed.

**Comment by wltb** [comment]
> Sadly, the node.js bundling software create a massive code. It is not only CodeMirror but Eslint & StyleLint that have to be packed.

Right. But all the way down, this [is discussed](https://codemirror.net/examples/bundle/#bundle-size). For minimizing, one can use
- minifiers (_babel_, _terse_),
- and _tree-shaking_.

According to the [rollup tutorial](https://rollupjs.org/tutorial/), it has a plugin for _terse_ which doesn't seem too hard to use on first glance, and tree-shaking is done automagically? I found [this link](https://www.codefeetime.com/post/tree-shaking-a-react-component-library-in-rollup/) where shaking is discussed a bit (in the bundlers configuration file, set the _sideEffects_ flag to **false**, or something. This [could work](https://codemirror.net/docs/guide/#functional-core%2C-imperative-shell)?!?).

**Comment by cyfung1031** [comment]
@erosman you can also consider monaco editor.

https://microsoft.github.io/monaco-editor/

unzip the zip file, copy the `min`  into extension folder.

But the size is around 3.5MB

**Comment by erosman** [comment]
> you can also consider monaco editor.

I use VS Code and the Monaco editor is a good editor.
I did ask them in [How to use Monaco Editor in a browser extension?](https://github.com/microsoft/monaco-editor/discussions/3908).

**Comment by DrakeTDL** [comment]
> > you can also consider monaco editor.
> 
> I use VS Code and the Monaco editor is a good editor. I did ask them in [How to use Monaco Editor in a browser extension?](https://github.com/microsoft/monaco-editor/discussions/3908).

You are able to use Monaco Editor in a browser extension (at least in Firefox) as I have modified FireMonkey to use it.

**Comment by erosman** [comment]
> You are able to use Monaco Editor in a browser extension (at least in Firefox) as I have modified FireMonkey to use it.

Have you replaced CodeMirror with Monaco and added ESLint & StyleLint, or are you using the built-in Monaco linter?

**Comment by cyfung1031** [comment]
> > you can also consider monaco editor.
> 
> I use VS Code and the Monaco editor is a good editor. I did ask them in [How to use Monaco Editor in a browser extension?](https://github.com/microsoft/monaco-editor/discussions/3908).

Monaco Editor can work like CM5 that just to put the script tag and then run the script. It will load the dependency automatically provided that you have the correct folder path.



<img width="602" alt="Screen Shot 2024-05-07 at 18 39 57" src="https://github.com/erosman/support/assets/44498510/97fb4d57-e240-462f-b85c-c0c62f5acd10">

These are the folder from the AMD bundled minimized version.

<img width="869" alt="Screen Shot 2024-05-07 at 18 40 45" src="https://github.com/erosman/support/assets/44498510/b9ba951f-fea9-44a1-b65b-0b1b27af7c75">


Put them into your extension's sub-folder.


Before adding the script, 
set the following on your page.

```
       const vsPath = "/public/lib/monaco-editor/min/vs";

        // Setting up Monaco Editor requirements
        let require = {
          paths: {
            vs: vsPath,
          },
        };

        window.require = (window.require || {});
        window.require.paths = (window.require.paths || {});
        Object.assign(window.require.paths, require.paths);
```

Without the `require.paths.vs`, the script do not know the folder path.

Then you can add the script with src=`${vsPath}/loader.js` to your `document.head`

Afterwards, a global object `monaco` will be created on the page.

These are the example codes for the setup.
```

    const editorOptions = {
      automaticLayout: true,
      foldingStrategy: 'indentation',
      lineNumbers: 'on',
      readOnly: false,
      minimap: {
        enabled: false,
      },
      cursorStyle: 'line',
      scrollBeyondLastLine: false,
      showUnused: true,
      showDeprecated: true,
    };

    const compilerOptions = {
      allowNonTsExtensions: true,
      checkJs: true,
      noImplicitAny: true,

      allowJs: true,
      noUnusedLocals: false,
      noFallthroughCasesInSwitch: false,
      noImplicitThis: false,

    };

      // global settings
      monaco.languages.typescript.javascriptDefaults.setCompilerOptions(Object.assign({
        target: monaco.languages.typescript.ScriptTarget.ES2018,
      }, compilerOptions));

      // when the editor is created
      monaco.editor.onDidCreateEditor(()=>{
         console.log('the monaco editor is created')
       })

      // the div will be the container of the editor.
      // the value is the input code of the editor. You can change it later too
      const editor = monaco.editor.create(div, Object.assign({
        value: '',
        language: 'javascript'
      }, editorOptions));

     // set the code
      editor.getModel().setValue(newCode);

    // set global theme (`vs` for light, `vs-dark` for dark)
       monaco.editor.setTheme("vs-dark");

    // detect code change
    editor.onDidChangeModelContent(()=>{
             console.log('code changed', editor.getValue())
    }

```

**Comment by DrakeTDL** [comment]
> Have you replaced CodeMirror with Monaco and added ESLint & StyleLint, or are you using the built-in Monaco linter?

From memory, I replaced most (if not all) instances of CM with ME and rework most features that were dependent on CM. As for the linter, I did use Monaco's built-in linter.

**Comment by erosman** [comment]
See also: [Editor & Linter](https://github.com/erosman/firemonkey/issues/3)
